server/src/main/java/org/opensearch/search/fetch/FetchPhase.java

```diff
@@ -196,7 +197,20 @@ public void execute(SearchContext context, String profileDescription) {
                         // for random access and don't optimize for sequential access - except for merging.
                         // So we do a little hack here and pretend we're going to do merges in order to
                         // get better sequential access.
-                        fieldReader = lf.getSequentialStoredFieldsReader()::document;
+                        StoredFieldsReader sequentialReader;
+                        // For scroll queries, try to get cached reader
+                        if (context.scrollContext() != null) {
+                            // To get a unique identifier for this segment to use as a cache key
+                            Object segmentKey = lf.getCoreCacheHelper() != null ? lf.getCoreCacheHelper().getKey() : currentReaderContext;
+                            sequentialReader = context.scrollContext().getCachedSequentialReader(segmentKey);
+                            if (sequentialReader == null) {
+                                sequentialReader = lf.getSequentialStoredFieldsReader();
+                                context.scrollContext().cacheSequentialReader(segmentKey, sequentialReader);
+                            }
+                        } else {
+                            sequentialReader = lf.getSequentialStoredFieldsReader();
+                        }
+                        fieldReader = sequentialReader::document;
```

Comment:
Slightly more readable?

```java
if (context.scrollContext() != null) {
    // To get a unique identifier for this segment to use as a cache key
    Object segmentKey = lf.getCoreCacheHelper() != null ? lf.getCoreCacheHelper().getKey() : currentReaderContext;
    sequentialReader = context.scrollContext().getCachedSequentialReader(segmentKey);
}
if (sequentialReader == null) {
    sequentialReader = lf.getSequentialStoredFieldsReader();
     if (context.scrollContext() != null) {
        context.scrollContext().cacheSequentialReader(segmentKey, sequentialReader);
     }
}
```

Reply:
Thanks Ankit In next commits. I have copy pasted as it is got some compile errors, will double check and optimize the code as required.